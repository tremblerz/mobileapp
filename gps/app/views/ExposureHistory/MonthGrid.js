import styled from '@emotion/native';
import dayjs from 'dayjs';
import localeData from 'dayjs/plugin/localeData';
import React from 'react';

dayjs.extend(localeData);

/**
 * Get an array of daily dates between start and end.
 *
 * @param {dayjs.Dayjs} start
 * @param {dayjs.Dayjs} end
 * @returns {dayjs.Dayjs[]}
 */
export function daysBetween(start, end, interval = 'day') {
  const days = [];
  let d = start;
  while (d.isBefore(end, interval) || d.isSame(end, interval)) {
    days.push(d);
    d = d.add(1, interval);
  }
  return days;
}

/**
 * Render a grid of days grouped by rows of n weeks.
 *
 * @param {{
 *   date?: dayjs.Dayjs,
 *   weeks?: number,
 *   renderDayHeader: (d: string) => import('react').ReactNode
 *   renderDay: (d: dayjs.Dayjs) => import('react').ReactNode
 * }} param0
 */
export const MonthGrid = ({
  date = dayjs(),
  weeks = 3,
  renderDayHeader,
  renderDay,
}) => {
  if (!renderDay || !renderDayHeader) {
    throw new Error('renderDay and renderDayHeader are required');
  }

  const headers = dayjs.localeData().weekdaysShort(); // 'mon', 'tue', ...

  const start = date
    .startOf('day')
    .startOf('week')
    .subtract(weeks - 1, 'week');

  const end = date.startOf('day').endOf('week');

  /** @type {dayjs.Dayjs[]} */
  const days = daysBetween(start, end, 'day');

  return (
    <Container>
      <HeaderRow>
        {headers.map(day => (
          <CellWrapper key={day}>{renderDayHeader(day)}</CellWrapper>
        ))}
      </HeaderRow>
      <DateGrid>
        {days.map(day => (
          <CellWrapper key={day.format()}>{renderDay(day)}</CellWrapper>
        ))}
      </DateGrid>
    </Container>
  );
};

const Container = styled.View`
  margin: 0 -8px; // offsets some of the percentage based cell padding
`;

const HeaderRow = styled.View`
  flex-direction: row;
`;

const DAYS_PER_WEEK = 7;

const CELL_WIDTH_PERCENT = (100 / DAYS_PER_WEEK).toFixed(0);

const DateGrid = styled.View`
  flex-direction: row;
  flex-wrap: wrap;
`;

const CellWrapper = styled.View`
  flex: 1;
  flex-basis: ${CELL_WIDTH_PERCENT}%;
  align-items: center;
  justify-content: center;
  min-height: 20px; // for the header text
`;
