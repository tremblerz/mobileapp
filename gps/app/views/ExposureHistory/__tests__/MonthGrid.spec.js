import { render } from '@testing-library/react-native';
import dayjs from 'dayjs';
import MockDate from 'mockdate';
import React from 'react';

import { MonthGrid } from '../MonthGrid';

const NOW = dayjs('2020-01-09T00:00:00-08:00').startOf('day');

const renderDayHeader = dayOfWeek => <span>{dayOfWeek}</span>;
const renderDay = date => <span>{date.format('D')}</span>;

const renderProps = { renderDayHeader, renderDay };

beforeEach(() => {
  MockDate.set(NOW.valueOf());
});

afterEach(() => {
  MockDate.reset();
});

it('renderDay is required', () => {
  jest.spyOn(console, 'error').mockImplementation();
  expect(() =>
    render(<MonthGrid renderDayHeader={renderDayHeader} />),
  ).toThrow();
});

it('renderDayHeader is required', () => {
  jest.spyOn(console, 'error').mockImplementation();
  expect(() => render(<MonthGrid renderDay={renderDay} />)).toThrow();
});

it('with no input date, defaults to 3 weeks ago from today', () => {
  const { asJSON } = render(<MonthGrid {...renderProps} />);

  expect(asJSON()).toMatchSnapshot();
});

it('with input date, defaults to 3 weeks ago from that date', () => {
  const date = dayjs('2019-12-02').subtract();
  const { asJSON } = render(<MonthGrid date={date} {...renderProps} />);

  expect(asJSON()).toMatchSnapshot();
});

it('allows specifying number of weeks', () => {
  const { asJSON } = render(<MonthGrid {...renderProps} weeks={1} />);

  expect(asJSON()).toMatchSnapshot();
});
