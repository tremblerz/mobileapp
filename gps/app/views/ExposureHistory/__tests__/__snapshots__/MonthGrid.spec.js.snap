// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allows specifying number of weeks 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "marginBottom": 0,
        "marginLeft": -8,
        "marginRight": -8,
        "marginTop": 0,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          Sun
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          Mon
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          Tue
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          Wed
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          Thu
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          Fri
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          Sat
        </span>
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "flexWrap": "wrap",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          5
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          6
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          7
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          8
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          9
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          10
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          11
        </span>
      </View>
    </View>
  </View>
</View>
`;

exports[`with input date, defaults to 3 weeks ago from that date 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "marginBottom": 0,
        "marginLeft": -8,
        "marginRight": -8,
        "marginTop": 0,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          Sun
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          Mon
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          Tue
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          Wed
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          Thu
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          Fri
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          Sat
        </span>
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "flexWrap": "wrap",
        }
      }
    />
  </View>
</View>
`;

exports[`with no input date, defaults to 3 weeks ago from today 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "marginBottom": 0,
        "marginLeft": -8,
        "marginRight": -8,
        "marginTop": 0,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          Sun
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          Mon
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          Tue
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          Wed
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          Thu
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          Fri
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          Sat
        </span>
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "flexWrap": "wrap",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          22
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          23
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          24
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          25
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          26
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          27
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          28
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          29
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          30
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          31
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          1
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          2
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          3
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          4
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          5
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          6
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          7
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          8
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          9
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          10
        </span>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexBasis": "14%",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "minHeight": 20,
          }
        }
      >
        <span>
          11
        </span>
      </View>
    </View>
  </View>
</View>
`;
